### ISSUE TITLE
Invalid type generated for Rect? from dart:ui package

### ISSUE THREAD

**Issue by sebastianbuechler** [dart-lang/mockito#662]
I have an interface for a share service like this:
```
import 'dart:ui';

abstract class ShareService {
  Future<void> share(
    String text, {
    String? subject,
    Rect? sharePositionOrigin,
  });
}
```

This generates the following mock with an `InvalideType` parameter:
```
class MockShareService extends _i1.Mock implements _i41.ShareService {
  @override
  _i27.Future<void> share(
    String? text, {
    String? subject,
    InvalidType sharePositionOrigin,
  }) =>
      (super.noSuchMethod(
        Invocation.method(
          #share,
          [text],
          {
            #subject: subject,
            #sharePositionOrigin: sharePositionOrigin,
          },
        ),
        returnValue: _i27.Future<void>.value(),
        returnValueForMissingStub: _i27.Future<void>.value(),
      ) as _i27.Future<void>);
}
```
Why can't mockito take this type (from a standard library) into the mock? How can I solve this?

**Comment by yanok** [comment]
same as https://github.com/dart-lang/mockito/issues/633?

**Comment by yanok** [comment]
* Mockito version?
* Does the generated file import `dart:ui`?
* What's the `which dart` output (assuming are use using Linux)? Is it the `dart` binary from the Flutter SDK? It looks like it's not.

**Comment by sebastianbuechler** [comment]
After some testing, I upgraded now to mockito 5.4.2 together with an upgrade to dart sdk 3.0.5. `which dart` shows "/c/src/flutter/bin/dart", I'm running it on windows with WSL and git bash.

With this setup I no longer get `InvalidType`, but `dynamic`, which seems to be quite similar as your mentioned #633 issue. And thus `dart:ui` is also not imported in the generated file.

Before, I used mockito `5.4.0` with flutter 3.10.5 if I remember correctly. Same setup for the dart installation.

**Comment by yanok** [comment]
well, change `InvalidType` -> `dynamic` is expected, but the real problem is the missing `dart:ui` import.

How do you run Mockito codegen? `dart run builder_runner build` or `flutter pub run build_runner build`? Could you please try both and check if there is any difference in the results?

**Comment by sebastianbuechler** [comment]
> well, change `InvalidType` -> `dynamic` is expected, but the real problem is the missing `dart:ui` import.
> 
> How do you run Mockito codegen? `dart run builder_runner build` or `flutter pub run build_runner build`? Could you please try both and check if there is any difference in the results?

I used to to run it with `flutter pub run build_runner build` but since there is now a warning that says we should use `dart run build_runner build` I use the dart version. I tried it with both and both do not import `dart:ui`. I also always use the flag `--delete-conflicting-outputs`.

**Comment by yanok** [comment]
I can't reproduce it.

Since you are using `dart run build_runner build`, could you please make sure `which dart` points to the same directory as `which flutter`? Or maybe there is another Dart SDK installed and `dart` is executed from there?

I don't have Windows machine at the moment, though I think WSL should work pretty much as a Linux... Could you maybe try if this is reproducible on Linux?

**Comment by yanok** [comment]
Uh, I see you already verified that `dart` binary comes from Flutter SDK, sorry for asking it again.

**Comment by yanok** [comment]
Turns out I can get a Windows instance pretty easily, but it can't run WSL...

I tried it on pure Windows, still can't reproduce. @sebastianbuechler Could you please try running it without WSL? (you would have to install Flutter SDK for Windows).

**Comment by yanok** [comment]
It would be also great if you could make a minimal repro package, upload it to Github and verify you can reproduce the issue by cloning it into a new directory and running `dart run build_runner build` there. Then, if it's still reproducible, maybe share `pubspec.lock` of this repro package. Otherwise I'm out of ideas.

**Comment by sebastianbuechler** [comment]
Thanks for looking into this. Not sure how turning off WSL would influence this behavior.

The change to `dynamic` is good enough for now as I don't see any issues now. Since nobody else is having this behavior I'll close the ticket.

**Comment by yanok** [comment]
Actually we have two other bugs that look super-similar (also not importing `dart:ui`). I'm also not sure about WSL, just trying to bisect it, and since I don't have access to WSL (but do have access to plain Windows), it would be great if you could test if you can reproduce with plain Windows setup.

But sure, I understand that `dynamic` might be good enough, so you don't see a point to invest more time in it.

**Comment by sebastianbuechler** [comment]
Found another occurrence of `InvalidType` from a dart:ui member: https://github.com/Stacked-Org/stacked/issues/977

**Comment by yanok** [comment]
Looking at the old issues found https://github.com/dart-lang/mockito/issues/579 that has the same problem. Turns out that's a bug in `package:build_resolvers` that was fixed in version 2.1.0.

@sebastianbuechler Could you please check `build_resolvers` version you are using and try to upgrade if it's below 2.1.0?

**Comment by sebastianbuechler** [comment]
**TLDR:** Check `pubspec.lock` file if you're using `build_resolvers` < 2.1.0 and if yes, try with removing it manually.

You're absolutely right, it's a weird issue since the reproduction fails all the time on so different issues. But the hint with the `build_resolvers` was the crucial point. As I don't have it in my direct dependencies I looked into my `pubspec.lock` file and found this older entry:
```
  build_resolvers:
    dependency: transitive
    description:
      name: build_resolvers
      sha256: "687cf90a3951affac1bd5f9ecb5e3e90b60487f3d9cdc359bb310f8876bb02a6"
      url: "https://pub.dev"
    source: hosted
    version: "2.0.10"
```

Even though I was using `build_runner: ^2.4.6` in my dev dependencies it did not update the `build_resolvers` version as they are referencing only `build_resolvers: ^2.0.0` (https://github.com/dart-lang/build/blob/3087308cd3649e83057155b1995a206527bb7028/build_runner/pubspec.yaml#L21C1-L21C26).

Now I deleted the entry from my `pubspec.lock` file and run again get packages. This resulted in the upgrade entry like this:
```
  build_resolvers:
    dependency: transitive
    description:
      name: build_resolvers
      sha256: "6c4dd11d05d056e76320b828a1db0fc01ccd376922526f8e9d6c796a5adbac20"
      url: "https://pub.dev"
    source: hosted
    version: "2.2.1"
```

Then the generation of the mocks went through without any problems. 

I'm not quite sure why my entry was not upgraded to the newest version automatically, but it could be that we used another package in the past that relied on the older version?

@yanok, thank you very much for solving this issue!

**Comment by yanok** [comment]
> Even though I was using `build_runner: ^2.4.6` in my dev dependencies it did not update the `build_resolvers` version as they are referencing only `build_resolvers: ^2.0.0`

Yes, that's what `^2.0.0` means: it's the same as `>=2.0.0 <3.0.0`, so pub was ok with the version `2.0.10`. You could upgrade it with `flutter pub upgrade build_resolvers` (And vice versa, a way to reproduce this behavior is to do `flutter pub downgrade build_resolvers` in the fresh package, and it will be downgraded to 2.0.10).

@natebosch I think we should make sure we don't run into this again (currently we have five different bugs with people reporting different symptoms of this), would be good to ban the know-bad version. What would be the idiomatic way of doing so? Would it be ok to just add `build_resolvers: ^2.1.0` to Mockito's `dev_dependencies`?

**Comment by natebosch** [comment]
Changing the dependency from `build_runner` wouldn't impact anyone who still has the old version of `build_runner`.

I'm not sure why so many flutter users are in the habit of very selectively upgrading immediate dependencies and leaving transitive dependencies at old versions. A `flutter pub upgrade` would have solved this, right?

> Now I deleted the entry from my `pubspec.lock` file and run again get packages.

Why did you do this instead of `flutter pub upgrade`? Are there docs suggesting users make manual edits in `pubspec.lock`?

**Comment by sebastianbuechler** [comment]
> Changing the dependency from `build_runner` wouldn't impact anyone who still has the old version of `build_runner`.
> 
> I'm not sure why so many flutter users are in the habit of very selectively upgrading immediate dependencies and leaving transitive dependencies at old versions. A `flutter pub upgrade` would have solved this, right?
> 
> > Now I deleted the entry from my `pubspec.lock` file and run again get packages.
> 
> Why did you do this instead of `flutter pub upgrade`? Are there docs suggesting users make manual edits in `pubspec.lock`?

It seems that `flutter pub upgrade` would have solved those issues, yes. 

I manually edit the lock file because I did not know the behavior of the upgrade command well enough, as I did not know that a) it was possible to update a single package and b) that it differs from manually upgrading the version in the `pubspec.yaml` file and running `pub get`. Additionally, I would never manually edit a lock file, if everything works as expected, and apparently this fixed the issue.

**Comment by yanok** [comment]
> Changing the dependency from `build_runner` wouldn't impact anyone who still has the old version of `build_runner`.

Right, but we would update Mockito's dependency on `builder_runner` at the same time. So we could just ask people to use the latest Mockito as a first step and hopefully that will solve some issues immediately.
 
> I'm not sure why so many flutter users are in the habit of very selectively upgrading immediate dependencies and leaving transitive dependencies at old versions. A `flutter pub upgrade` would have solved this, right?

I'm not a Flutter user and I do most of my stuff in monorepo anyway, but if I touch pub packages, I don't always do `dart pub upgrade` either. I think people tend to just bump their immediate dependencies if needed and then defer to pub to find a minimal viable upgrade. So there are less things to rebuild. I thought that's a pub feature ;)

**Comment by natebosch** [comment]
> Right, but we would update Mockito's dependency on `builder_runner` at the same time.

The mockito `dev_dependencies` do not impact downstream consumers, so updating to the latest mockito would not force the latest `build_runner`.

> So there are less things to rebuild

AFAIK once the pub solve changes any build caches we have across dart of flutter will be invalidated. It's possible flutter has some more fine grained invalidation, but I would be very surprised.

**Comment by yanok** [comment]
> The mockito `dev_dependencies` do not impact downstream consumers, so updating to the latest mockito would not force the latest `build_runner`.

Oh, right. I forgot that people using Mockito have to add `build_runner` as their dev dependency directly.
